stdlib/InteractiveUtils/src/editless.jl

```diff
@@ -5,6 +5,165 @@
 import Base.shell_split
 using Base: find_source_file
 
+struct EditorEntry{P,Fn}
+  pattern::P
+  fn::Fn
+  background::Bool
+  takesline::Bool
+  priority::Float64
+end
+const editors = Vector{EditorEntry}(undef,0)
+
+struct EditorCommand{C,E}
+    parsed_command::C
+    entry::E
+end
+function parse_command(entry::EditorEntry,command,path,line)
+    if entry.takesline
+        cmd = entry.fn(command,path,line)
+        if !(cmd isa Nothing)
+            return EditorCommand(cmd,entry), true
+        end
+    end
+
+    cmd = entry.fn(command,path)
+    if !(cmd isa Nothing)
+        return EditorCommand(cmd,entry), false
+    end
+
+    nothing, false
+end
+
+function Base.run(x::EditorCommand{Cmd})
+    if x.entry.background
+        run(pipeline(x.parsed_command, stderr=stderr), wait=false)
+    else
+        run(x.parsed_command)
+    end
+end
+Base.run(x::EditorCommand{Function}) = x.parsed_command()
+
+"""
+    define_editor(fn,pattern;background=true,priority=0)
+
+Define a new editor matching `pattern` that can be used to open a file
+(possibly at a given line number) using `fn`.
+
+The `fn` argument is a function that determines how to open a file with the
+given editor.  It should take 2 or 3 arguments, as follows:
+
+* `command` - an array of strings representing the editor command. It can be
+  safely interpolated into a command created using backtick notation.
+* `path`  - the path to the source file to open
+* `line` - the optional line number to open to; if specified the returned command
+  must open the file at the given line.
+
+`fn` must return either an appropriate `Cmd` object to open a
+file, a function (taking 0 arguments) that will open the file directly
+(returning a `Cmd` is the preferred approach), or `nothing`. Use `nothing`
+to indicate that this editor is not appropriate for the current environment
+and another editor should be attempted.
+
+The `pattern` argument is string, regular expression, or an array of strings and
+regular expressions. For the `fn` to be called one of the patterns must match
+the value of `EDITOR`, `VISUAL` or `JULIA_EDITOR`.  For strings, only whole
+words can match (i.e. "vi" doesn't match "vim -g" but will match "/usr/bin/vi
+-m").
+
+If multiple defined editors match, the one with the highest priority is selected,
+and if there is a tie in priority, the first editor added will be used.
+
+By default the editor is opened in the background, but terminal-based editors
+will want to set `background=false`.
```

Comment:
This is not 100% clear to me—does this indicate whether the command will block until the editing is done?